### PR TITLE
New package charm

### PR DESCRIPTION
**Pull request by eschnett**
Charm++ is a parallel programming framework in C++ supported by
    an adaptive runtime system, which enhances user productivity and
    allows programs to run portably from small multicore computers
    (your laptop) to the largest supercomputers.

### CODE REVIEW

File: var/spack/repos/builtin/packages/charm/package.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    variant("mpi", default=False,
+            description="Use MPI as communication mechanism")
+    variant("multicore", default=False,
+            description="Disable inter-node communication")
+    variant("net", default=False,
+            description="Use net communication mechanism")
+    variant("netlrts", default=True,
+            description="Use netlrts communication mechanism")
+    variant("verbs", default=False,
+            description="Use Infinitband as communication mechanism")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/charm/package.py]: Typo: infintband -> infiniband

File: var/spack/repos/builtin/packages/charm/package.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            comm = "multicore"
+        if "+net" in spec:
+            comm = "net"
+        if "+netlrts" in spec:
+            comm = "netlrts"
+        if "+verbs" in spec:
+            comm = "verbs"
+
+        plat = sys.platform
+        if plat.startswith("linux"): plat = "linux"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/charm/package.py]: Flake8 doesn't like this. Can you split this into 2 lines?

**Comment by eschnett** [comment]
Corrected typo and flake8 error.

**Comment by philmiller-charmworks** [comment]
The 'gcc' option is probably not quite the right thing. I think that compiler option file in our tree will actually default away from the desired compiler selection, and will lose options that we want to pass to the compiler based on what's in the usual configuration files when not passing the 'gcc' option.

We have some modifications to Charm++ on a branch to make its build scripts more friendly to Spack that we're working on reviewing and integrating soon.

**Comment by eschnett** [comment]
@philmiller-charmworks Are you suggesting to (1) simply omit the `gcc` option, or (2) actually detect the compiler vendor and pass it to Charm++?

**Comment by tgamblin** [comment]
@philmiller-charmworks: what's the timeline on the build changes?  And if those are going to be in some future charm++ version, does it make sense to merge these changes for the current charm++ version, and update things once you have the new options?

**Comment by tgamblin** [comment]
@philmiller-charmworks: ping

**Comment by tgamblin** [comment]
I'm going to go ahead and merge this one as-is, since @eschnett has provided it and it looks reasonable.  We can add updates as they become available.

**Comment by tgamblin** [comment]
@bhatele @philmiller-charmworks: please feel free to submit new PRs on top of this one.

**Pull request merged**